# Spotify Recommendation Engine

For this capstone project, I will be attempting to use classification models to try to predict and recommend songs that I would enjoy, based on my own Spotify data. All of the data used in this capstone was collected by me using the Spotify API, Spotipy.

In [1]:
import pandas as pd
import numpy as np
import yaml
from spotipy.oauth2 import SpotifyOAuth
import json
import re
import sys
import itertools
import seaborn as sns
from tabulate import tabulate

# import matplotlib.pyplot as plt
# import matplotlib.ticker as ticker

from sklearn import datasets, linear_model
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.preprocessing import MinMaxScaler
from sklearn.decomposition import PCA
from sklearn.manifold import TSNE
from sklearn.model_selection import KFold
from sklearn.metrics import accuracy_score
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression



from sklearn.feature_selection import RFECV
from sklearn.model_selection import train_test_split, StratifiedKFold, GridSearchCV, RandomizedSearchCV
from sklearn.metrics import classification_report


# import spotipy
# from spotipy.oauth2 import SpotifyClientCredentials
# from spotipy.oauth2 import SpotifyOAuth
# import spotipy.util as util

import warnings
warnings.filterwarnings("ignore")

pd.set_option("max_rows", None)
pd.set_option('display.max_columns', None)

In [2]:
#loading datasets

#Datasets used to rate the training dataset
top_artist_df = pd.read_pickle("top_artists.pkl")
top_tracks_df = pd.read_pickle("top_tracks.pkl")

#Training Dataset
saved_tracks_df = pd.read_pickle("saved_tracks.pkl")

#Validation dataset
recommendation_tracks_df = pd.read_pickle("recommendation_tracks.pkl")

In [3]:
#removing duplicates of songs from the validation playlist
recommendation_tracks_df = recommendation_tracks_df.drop_duplicates(subset='id', keep="first").reset_index()

First thing I want to do is to check the shapes and variables in my playlists.

In [4]:
top_artist_df.shape

(60, 6)

In [5]:
top_artist_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 60 entries, 0 to 59
Data columns (total 6 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   id         60 non-null     object
 1   uri        60 non-null     object
 2   type       60 non-null     object
 3   name       60 non-null     object
 4   genres     60 non-null     object
 5   followers  60 non-null     int64 
dtypes: int64(1), object(5)
memory usage: 2.9+ KB


In [6]:
#Changing "id" to "artist_id" to match other dataset
top_artist_df = top_artist_df.rename(columns={'id': 'artist_id'})
top_artist_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 60 entries, 0 to 59
Data columns (total 6 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   artist_id  60 non-null     object
 1   uri        60 non-null     object
 2   type       60 non-null     object
 3   name       60 non-null     object
 4   genres     60 non-null     object
 5   followers  60 non-null     int64 
dtypes: int64(1), object(5)
memory usage: 2.9+ KB


In [7]:
#Looking at my top artists, which will be important for applying a ranking to the main dataset used for training
top_artist_df.head(20)

artist_id                                    uri    type  \
0   3HiuzBlSW7pGDXlSFMhO2g  spotify:artist:3HiuzBlSW7pGDXlSFMhO2g  artist   
1   2Waw2sSbqvAwK8NwACNjVo  spotify:artist:2Waw2sSbqvAwK8NwACNjVo  artist   
2   2GeutBKR3pXPk1OBmDYWD3  spotify:artist:2GeutBKR3pXPk1OBmDYWD3  artist   
3   1yxSLGMDHlW21z4YXirZDS  spotify:artist:1yxSLGMDHlW21z4YXirZDS  artist   
4   3WrFJ7ztbogyGnTHbHJFl2  spotify:artist:3WrFJ7ztbogyGnTHbHJFl2  artist   
5   2DaxqgrOhkeH0fpeiQq2f4  spotify:artist:2DaxqgrOhkeH0fpeiQq2f4  artist   
6   3TVXtAsR1Inumwj472S9r4  spotify:artist:3TVXtAsR1Inumwj472S9r4  artist   
7   0RESbWvOMyua0yuyVrztJ5  spotify:artist:0RESbWvOMyua0yuyVrztJ5  artist   
8   13y7CgLHjMVRMDqxdx0Xdo  spotify:artist:13y7CgLHjMVRMDqxdx0Xdo  artist   
9   5K4W6rqBFWDnAN6FQUkS6x  spotify:artist:5K4W6rqBFWDnAN6FQUkS6x  artist   
10  0BEYTctVmnYa5yStp4Jpab  spotify:artist:0BEYTctVmnYa5yStp4Jpab  artist   
11  5f7VJjfbwm532GiveGC0ZK  spotify:artist:5f7VJjfbwm532GiveGC0ZK  artist   
12  67nwj3Y5sZQLl72VNUHEYE  spotify:artist:67nwj3Y5sZQLl72VNUHEYE  artist   
13  4cUkGQyhLFqKHBtL58HYVp  spotify:artist:4cUkGQyhLFqKHBtL58HYVp  artist   
14  6l3HvQ5sa6mXTsMTB19rO5  spotify:artist:6l3HvQ5sa6mXTsMTB19rO5  artist   
15  2d0hyoQ5ynDBnkvAbJKORj  spotify:artist:2d0hyoQ5ynDBnkvAbJKORj  artist   
16  2p1fiYHYiXz9qi0JJyxBzN  spotify:artist:2p1fiYHYiXz9qi0JJyxBzN  artist   
17  3tJoFztHeIJkJWMrx0td2f  spotify:artist:3tJoFztHeIJkJWMrx0td2f  artist   
18  6oMuImdp5ZcFhWP0ESe6mG  spotify:artist:6oMuImdp5ZcFhWP0ESe6mG  artist   
19  738wLrAtLtCtFOLvQBXOXp  spotify:artist:738wLrAtLtCtFOLvQBXOXp  artist   

                        name  \
0                Young Dolph   
1                 Bo Burnham   
2             Laurent Korcia   
3            Black Eyed Peas   
4                The Beatles   
5                      Oasis   
6                      Drake   
7                  Key Glock   
8                 Gucci Mane   
9                 Kanye West   
10                 Destructo   
11                  Lil Baby   
12                      Wale   
13              Channel Tres   
14                   J. Cole   
15  Rage Against The Machine   
16                    Skepta   
17              Moneybagg Yo   
18                     Migos   
19               Major Lazer   

                                               genres  followers  
0   [gangster rap, hip hop, memphis hip hop, rap, ...    1831256  
1                                             [comic]    1297100  
2                                            [violin]       1497  
3                           [dance pop, pop, pop rap]    6276967  
4   [beatlesque, british invasion, classic rock, m...   21322261  
5   [beatlesque, britpop, madchester, permanent wa...    6847875  
6   [canadian hip hop, canadian pop, hip hop, rap,...   59353055  
7   [hip hop, memphis hip hop, rap, southern hip h...    1023419  
8   [atl hip hop, dirty south rap, gangster rap, h...    4082289  
9                                  [chicago rap, rap]   15846203  
10  [bass house, edm, electro house, electronic trap]      56604  
11                 [atl hip hop, atl trap, rap, trap]    8946829  
12  [hip hop, pop, pop rap, r&b, rap, southern hip...    3393642  
13                               [indie soul, la pop]     179207  
14  [conscious hip hop, hip hop, north carolina hi...   15222013  
15  [alternative metal, alternative rock, consciou...    4592729  
16  [birmingham grime, grime, instrumental grime, ...     985200  
17  [hip hop, memphis hip hop, rap, southern hip h...    2508941  
18                  [atl hip hop, pop rap, rap, trap]   11963671  
19  [dance pop, edm, electro house, moombahton, po...    6250741

In [8]:
top_tracks_df.shape

(60, 35)

In [9]:
# Checking all columns look consistent
top_tracks_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 60 entries, 0 to 59
Data columns (total 35 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   id                  60 non-null     object 
 1   name                60 non-null     object 
 2   popularity          60 non-null     int64  
 3   type                60 non-null     object 
 4   is_local            60 non-null     bool   
 5   explicit            60 non-null     bool   
 6   duration_ms         60 non-null     int64  
 7   disc_number         60 non-null     int64  
 8   track_number        60 non-null     int64  
 9   artist_id           60 non-null     object 
 10  artist_name         60 non-null     object 
 11  album_artist_id     60 non-null     object 
 12  album_artist_name   60 non-null     object 
 13  album_id            60 non-null     object 
 14  album_name          60 non-null     object 
 15  album_release_date  60 non-null     object 
 16  album_trac

In [10]:
top_tracks_df.isnull().sum()

id                    0
name                  0
popularity            0
type                  0
is_local              0
explicit              0
duration_ms           0
disc_number           0
track_number          0
artist_id             0
artist_name           0
album_artist_id       0
album_artist_name     0
album_id              0
album_name            0
album_release_date    0
album_tracks          0
album_type            0
genres                0
album_genres          0
danceability          0
energy                0
key                   0
loudness              0
mode                  0
speechiness           0
acousticness          0
instrumentalness      0
liveness              0
valence               0
tempo                 0
uri                   0
track_href            0
analysis_url          0
time_signature        0
dtype: int64

In [11]:
saved_tracks_df.shape

(1133, 36)

In [12]:
saved_tracks_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1133 entries, 0 to 1132
Data columns (total 36 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   id                  1131 non-null   object 
 1   name                1133 non-null   object 
 2   popularity          1133 non-null   int64  
 3   type                1131 non-null   object 
 4   is_local            1133 non-null   bool   
 5   explicit            1133 non-null   bool   
 6   duration_ms         1131 non-null   float64
 7   disc_number         1133 non-null   int64  
 8   track_number        1133 non-null   int64  
 9   artist_id           1133 non-null   object 
 10  artist_name         1133 non-null   object 
 11  album_artist_id     1133 non-null   object 
 12  album_artist_name   1133 non-null   object 
 13  album_id            1133 non-null   object 
 14  album_name          1133 non-null   object 
 15  album_release_date  1133 non-null   object 
 16  album_

In [13]:
saved_tracks_df.isnull().sum()

id                    2
name                  0
popularity            0
type                  2
is_local              0
explicit              0
duration_ms           2
disc_number           0
track_number          0
artist_id             0
artist_name           0
album_artist_id       0
album_artist_name     0
album_id              0
album_name            0
album_release_date    0
album_tracks          0
album_type            0
added_at              0
genres                0
album_genres          0
danceability          2
energy                2
key                   2
loudness              2
mode                  2
speechiness           2
acousticness          2
instrumentalness      2
liveness              2
valence               2
tempo                 2
uri                   2
track_href            2
analysis_url          2
time_signature        2
dtype: int64

This dataset seems to be missing data for 2 rows of data in a dataset of 1100+. I think I can safely drop these rows without effecting data.

In [14]:
#dropping na values
saved_tracks_df = saved_tracks_df.dropna()
saved_tracks_df.isnull().sum()

#looking good now no missing data

id                    0
name                  0
popularity            0
type                  0
is_local              0
explicit              0
duration_ms           0
disc_number           0
track_number          0
artist_id             0
artist_name           0
album_artist_id       0
album_artist_name     0
album_id              0
album_name            0
album_release_date    0
album_tracks          0
album_type            0
added_at              0
genres                0
album_genres          0
danceability          0
energy                0
key                   0
loudness              0
mode                  0
speechiness           0
acousticness          0
instrumentalness      0
liveness              0
valence               0
tempo                 0
uri                   0
track_href            0
analysis_url          0
time_signature        0
dtype: int64

In [15]:
#saved_tracks_df.info()

In [16]:
#renaming main playlist to df for simplicity
df = saved_tracks_df

First thing I want to do is apply a rating to my saved songs dataset. To apply a rating, I will give the songs that are in my "top artists" a rating of 1, and everything else a 0. "Rating" will be the variable I will attempt to predict based on the content of my songs. 

I originally wanted to use specific song data to train and test models, but the "top tracks" playlist was only 60 songs long, so I was only able to rate a maximum of 60 songs. Whereas when using top artist data, I was able to apply a rating to 399 songs. I also noticed a large portion of my top songs were by my top artists, so there was a large overlap. I am confident I can get good results by analyzing this song data from my top artists.

In [17]:
#Only 20 top songs get rated when using top songs, 
#think it best to use a rating system with my top artists as that gives me 399 ratings


#saved_tracks_df['ratings'] = saved_tracks_df['id'].apply(lambda x: 1 if x in top_tracks_df.values else 0)
#saved_tracks_df['ratings'].head(25)

In [18]:
#Lambda function to add a rating for all songs that are by one of my top artists

df['ratings'] = df['artist_id'].apply(lambda x: 1 if x in top_artist_df.values else 0)
df['ratings']

0       0
1       1
2       0
3       0
4       0
5       0
6       0
7       0
8       1
9       1
10      1
11      1
12      1
13      1
14      1
15      1
16      0
17      0
18      1
19      1
20      1
21      1
22      1
23      1
24      0
25      0
26      1
27      0
28      0
29      1
30      0
31      1
32      0
33      0
34      0
35      0
36      0
37      0
38      0
39      0
40      0
41      0
42      0
43      0
44      1
45      1
46      1
47      1
48      1
49      0
50      0
51      1
52      1
53      0
54      1
55      1
56      0
57      1
58      1
59      0
60      0
61      1
62      0
63      1
64      1
65      0
66      1
67      1
68      0
69      0
70      0
71      0
72      0
73      0
74      0
75      0
76      0
77      1
78      0
79      0
80      0
81      0
82      0
83      0
84      0
85      0
86      0
87      0
88      0
89      0
90      0
91      1
92      0
93      1
94      0
95      0
96      1
97      0
98      1
99      0


In [19]:
#viewing df to see ratings
df.head(25)

id                                               name  \
0   6xgwv5ao2qYqmpkuG1etFc                                        Minor Waltz   
1   3CFR1oY47Ph2fULELeH6Rd                             Schindler's List Theme   
2   2HCs8VTLo6Vkzu06gVt1C1                                               瘋狂世界   
3   5B0qLRZewG1huy6o2JgQAo                                                 交換   
4   0vDU9V9SLZgqkQJ68HhnEU                                            Element   
5   6Gifhuvyk4uPDGSRrtw8Y6  Schwanengesang, D. 957: Ständchen "Leise flehe...   
6   0vf2eBw2inhl8y61cYQMv2                                         Devil Town   
7   6K0LB4XItoRSkKZSkVTS4d                                    Substance Abuse   
8   3TXy6nchgKeYlVwOKNk9lE                                          100 Shots   
9   2MEhck5D2vq7kPRkfJ93HI                                     Everybody Know   
10  26v2Gf4Af7Fav0jTQluetL                                           Penguins   
11  5YGhm6pI5s0uxJ2DwrDYqO                            Water on Water on Water   
12  2bGro4nCmkUx2xdEjcOTU0                                          Cray Cray   
13  1Lyj66RLoLz4QRoecQruMk                                         Reflection   
14  2G6npEiE9YZHdvkKP7YSd3                            1 Scale (feat. G Herbo)   
15  047IfBzxGgHyPpQTwMLSlP                                             Summo'   
16  1H5aqJh1n2xJPTVc7D0aqR                                          Alter Ego   
17  6mQOJJxlDLifWgg48jlU7A              Tapped (feat. Channel Tres & Celeste)   
18  2sRH0V7gWZ9p3QHfqhhOk8                              Sexy Black Timberlake   
19  3fOObCPvmiAbySTOvYucqF                                          Raw Power   
20  2GQEM9JuHu30sGFvRYeCxz                                              Faded   
21  36LQYp5FNsuT5voIO9bdfZ                                   Came For The Low   
22  5QUbOFOkbDlt4ZY7bdL4Am                        I Admit It (feat. 24kGoldn)   
23  0Kuq3DdX1sPGpJsAxeBsxx                                   How Does It Feel   
24  3M6xaTs8eskSIBEa8CA4Bg                                        READY TO GO   

    popularity            type  is_local  explicit  duration_ms  disc_number  \
0            6  audio_features     False     False     228893.0            1   
1            1  audio_features     False     False     225373.0            1   
2            8  audio_features     False     False     170160.0            1   
3            2  audio_features     False     False     195507.0            1   
4           32  audio_features     False      True     131056.0            1   
5           52  audio_features     False     False     260000.0            1   
6           82  audio_features     False     False     179955.0            1   
7           13  audio_features     False      True     212245.0            1   
8           73  audio_features     False      True     264819.0            1   
9           54  audio_features     False      True     147173.0            1   
10          66  audio_features     False      True     238013.0            1   
11          67  audio_features     False      True     223520.0            1   
12          42  audio_features     False      True     145107.0            1   
13          47  audio_features     False      True     204493.0            1   
14          45  audio_features     False      True     176040.0            1   
15          55  audio_features     False      True     156147.0            1   
16          53  audio_features     False      True     205497.0            1   
17          45  audio_features     False      True     188398.0            1   
18          53  audio_features     False      True     223103.0            1   
19          38  audio_features     False      True     270221.0            1   
20          64  audio_features     False     False     223480.0            1   
21          59  audio_features     False     False     195976.0            1   
22          62  audio_features     False      True     173951.0            1 

In [20]:
df.ratings.sum()

399

Now That I have ratings to my song data, I will move right into training my data.

For training the data, I am looking only at the variables that describe the musical contents of the song. I am not interested in popularity, song length, whether song is explicit, or the song type.

### Training Data

In [21]:
# Training data
X = df[['danceability', 'energy', 'key', 'loudness', 'mode', 'speechiness', 'acousticness', 'instrumentalness', 
        'liveness', 'valence', 'tempo', 'time_signature', 'genres']]
y = df['ratings']


# Create genre columns to turn the genres column into a numerical column
X = X.drop('genres', 1).join(X['genres'].str.join('|').str.get_dummies())

#Loading in copy of recommendation tracks from spotify's own algorythms, will use this later for validation
X_recommend = recommendation_tracks_df.copy()

# Create genre columns to turn the genres column into a numerical column
X_recommend = X_recommend.drop('genres', 1).join(X_recommend['genres'].str.join('|').str.get_dummies())


#Ensuring features are consistent
X = X[X.columns.intersection(X_recommend.columns)]
X_recommend = X_recommend[X_recommend.columns.intersection(X.columns)]


# Train-test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state= 7)

In [22]:
#checking to make sure shape matches up
X_train.shape

(904, 263)

In [23]:
y_train.shape

(904,)

In [24]:
X_recommend.shape

(4338, 263)

### Feature Importance

In [25]:
#Cross Validation for features

#rf = RandomForestClassifier(n_estimators = 1000, random_state = 7 )
#rfecv = RFECV(estimator=rf, step=1, n_jobs=-1, cv=StratifiedKFold(2), verbose=1, scoring='roc_auc')
#rfecv.fit(X_train, y_train)

#print(f'Optimal number of features: {rfecv.n_features_}')

#plt.figure()
#plt.xlabel('Number of features selected')
#plt.ylabel('Cross validation score')
#plt.plot(range(1, len(rfecv.grid_scores_)+1), rfecv.grid_scores_)
#plt.show()

#Optimal number of features is 232

#commented out to save run time when restarting kernel

In [26]:
#Random Forest classifier
rf = RandomForestClassifier(n_estimators = 1000, random_state = 7 )

#fitting model now to see a list of important features for my song data
rf.fit(X_train, y_train)

important_features = {}
for x,y in enumerate(rf.feature_importances_):
    important_features[x] = y

important_features_list = sorted(important_features,
                                 key=important_features.get,
                                 reverse=True)

# Printing the top features
for i in range(len(important_features_list[:50])):
    print(f"{i+1}. {X_train.columns[important_features_list[i]]} {important_features[important_features_list[i]]}")

1. rap 0.07338067434111464
2. speechiness 0.050407626497393415
3. energy 0.04533777337736288
4. danceability 0.043416393481863165
5. instrumentalness 0.042697105051097604
6. loudness 0.04092880252721668
7. acousticness 0.03507823743659393
8. detroit hip hop 0.034302485156577356
9. tempo 0.03183338031885002
10. valence 0.03137924787590353
11. liveness 0.030653429870509764
12. hip hop 0.027732750028463874
13. chicago rap 0.024633724420120705
14. new orleans rap 0.021977039079103367
15. atl hip hop 0.02195072244400758
16. underground hip hop 0.021663480368939565
17. escape room 0.019497149952502593
18. west coast rap 0.01829968900060041
19. key 0.018220628076729473
20. kansas city hip hop 0.01813147225963202
21. psychedelic hip hop 0.017594989537704227
22. east coast hip hop 0.017516561768414304
23. pop rap 0.01721553619200151
24. canadian hip hop 0.01696395916559439
25. conscious hip hop 0.016271585267872674
26. southern hip hop 0.0150330855380227
27. melodic rap 0.013817814133315462
28.

This lists of feature importance at first glance looks accurate. My favorite type of music, as well as my favorite/ most listened to artists I would classify as rap or hip hop. Most of the music I listen to is at the gym, where I prefer loud, fast-tempo rap songs to fuel my workout.

One other thing I notice, is that the music type is not extremely diverse. Most of the genres of importance could theoretically all fall under the same large umbrella of "rap music". (several different types of hip hop and rap). My music taste is very diverse, as seen in my saved_playlist dataframe. I have different types of musical tastes including classical, rock, alternative, even some songs sung in Chinese from artists who I cannot pronounce the name of, from the 1940's. Based on my top artist data, I would predict that the models will recommend primarily rap and hip hop type music. While not as diverse as it could be, I believe the results I get may be even more accurate in that the songs it recommends will have a higher chance of me liking them.

The 3 types of classification models I want to test my on data on are Logistical Regression, kNN, and Random Forest. Between these 3 models, I will select the best model and test it on a dataset of songs from someone else randomly found on Kaggle, and try to predict which songs from that dataset I will like the most.

### Logistic Regression

In [27]:
#Logistic Regression
lr = LogisticRegression()

#fit model with train data
lr.fit(X_train, y_train)

print('The accuracy score is: ' + str(accuracy_score(lr.predict(X_test), y_test)))
print(classification_report(lr.predict(X_test), y_test))

The accuracy score is: 0.8986784140969163
              precision    recall  f1-score   support

           0       0.95      0.91      0.93       158
           1       0.80      0.88      0.84        69

    accuracy                           0.90       227
   macro avg       0.87      0.89      0.88       227
weighted avg       0.90      0.90      0.90       227



In [28]:
#Checking accuracy after SCALING Data
minmax_scaler = MinMaxScaler()
X_train_scaled = pd.DataFrame(minmax_scaler.fit_transform(X_train), columns=X_train.columns)

#Logistic Regression
lr = LogisticRegression()

#fit model with train data
lr.fit(X_train_scaled, y_train)

print('The accuracy score is: ' + str(accuracy_score(lr.predict(X_test), y_test)))
print(classification_report(lr.predict(X_test), y_test))

The accuracy score is: 0.6651982378854625
              precision    recall  f1-score   support

           0       1.00      0.67      0.80       227
           1       0.00      0.00      0.00         0

    accuracy                           0.67       227
   macro avg       0.50      0.33      0.40       227
weighted avg       1.00      0.67      0.80       227



It seems that scaling the data makes the accuracy and models perform much worse. I will test my data on more models to see how this compares before making any decisions.

### KNN

In [29]:
knn = KNeighborsClassifier()

#First thing I want to do is test for the optimal number of n_neighbors
knn_params = {'n_neighbors': range(1, 20)}

knn_gcv = GridSearchCV(knn, knn_params, n_jobs=1, cv=StratifiedKFold(2), scoring='roc_auc')
knn_gcv.fit(X_train, y_train)
knn_gcv.best_params_, knn_gcv.best_score_

({'n_neighbors': 2}, 0.585026889408075)

In [30]:
#KNN model with training data, using optimal n_neighbors of 2
knn = KNeighborsClassifier(n_neighbors = 2)

#Fitting Model with UNSCALED training data
knn.fit(X_train, y_train)

#Accuracy score and classification report
print('The accuracy score is: ' + str(accuracy_score(knn.predict(X_test), y_test)))
print(classification_report(knn.predict(X_test), y_test))

The accuracy score is: 0.6563876651982379
              precision    recall  f1-score   support

           0       0.85      0.70      0.77       185
           1       0.26      0.48      0.34        42

    accuracy                           0.66       227
   macro avg       0.56      0.59      0.55       227
weighted avg       0.74      0.66      0.69       227



In [31]:
#Checking Model after SCALING Data
#KNN model with training data, using optimal n_neighbors of 2
knn = KNeighborsClassifier(n_neighbors = 2)

#Fitting Model with training data
knn.fit(X_train_scaled, y_train)

#Accuracy score and classification report
print('The accuracy score is: ' + str(accuracy_score(knn.predict(X_test), y_test)))
print(classification_report(knn.predict(X_test), y_test))

The accuracy score is: 0.6651982378854625
              precision    recall  f1-score   support

           0       1.00      0.67      0.80       227
           1       0.00      0.00      0.00         0

    accuracy                           0.67       227
   macro avg       0.50      0.33      0.40       227
weighted avg       1.00      0.67      0.80       227



Using the optimal n_neighbors of 2 (as opposed to the default n_neighbors of 5) improved the accuracy score about 0.05 from roughly 0.60 to 0.65. Then a very minor inmprovement of 0.01 after using scaled data. This is ok, but does not compare to the Logistic Regression Model. Think it is sae to say this is the least effective

The final model I will test my data on will be a Random Forest Model.

### Random Forest

In [32]:
rfc = RandomForestClassifier()

#First thing I want to do is test for the optimal parameters
rfc_params = {'min_samples_leaf': [1, 3, 5, 10, 20], 
              'max_depth': [2, 4, 5, 10, 12, 16, 20], 
             }

rfc_gcv = GridSearchCV(rfc, rfc_params, n_jobs=1, cv=StratifiedKFold(3), scoring='roc_auc')
rfc_gcv.fit(X_train, y_train)
rfc_gcv.best_params_, rfc_gcv.best_score_

({'max_depth': 20, 'min_samples_leaf': 1}, 0.9750416916512732)

In [33]:
#Finally, will test data on a Random Forest Classifier
rfc = RandomForestClassifier(max_depth = 20, min_samples_leaf = 1)

#Fitting Model with training data
rfc.fit(X_train, y_train)

#Accuracy score and classification report
print('The accuracy score is: ' + str(accuracy_score(rfc.predict(X_test), y_test)))
print(classification_report(rfc.predict(X_test), y_test))

The accuracy score is: 0.9383259911894273
              precision    recall  f1-score   support

           0       0.97      0.94      0.95       157
           1       0.87      0.94      0.90        70

    accuracy                           0.94       227
   macro avg       0.92      0.94      0.93       227
weighted avg       0.94      0.94      0.94       227



In [34]:
#Finally, will test SCALED data on a Random Forest Classifier
#rfc = RandomForestClassifier(max_depth = 16, min_samples_leaf = 1)

#Fitting Model with training data
#rfc.fit(X_train_scaled, y_train)

#Accuracy score and classification report
#print('The accuracy score is: ' + str(accuracy_score(rfc.predict(X_test), y_test)))
#print(classification_report(rfc.predict(X_test), y_test))


#Random Forest Does not need scaled data

This Random Forest Classifier Model seems to perform the best over any model, scaled or unscaled. This will be the model I use.

### Recommendations

For the recommendations, I will be importing a dataset I collected from my spotify of spotify's recommended tracks that they believe I would enjoy. This will be the validation set that I will use to see which of those songs my model thinks I will enjoy the most

In [35]:
#creating a rating column for recommendation playlist
recommendation_tracks_df['ratings'] = rfc_gcv.predict(X_recommend)

#Creating a probability rating for the songs
recommendation_tracks_df['prob_ratings'] = rfc_gcv.predict_proba(X_recommend)[:,1]

In [36]:
#Sort playlist by ratings
recommendation_tracks_df[recommendation_tracks_df['ratings'] == 1].sort_values('prob_ratings', ascending = False).head(50)

index                      id  \
3121   3725  3aGibUHhQyBsyumYHylw0K   
2966   3474  6HZILIRieu8S0iqY8kIKhj   
4010   5091  2UciqgwVDwro0fev3MFbxJ   
2976   3489  561jH07mF1jHuk7KlaeF0s   
3309   4001  0y9uTzK9cNKSAEHnpeRG8C   
2191   2460  28FGV3ORH14MYORd7s5dlU   
2200   2473  40YbWniIEmqy6s58fYXLUh   
3218   3873  3wScL5W8H40zzCKN0atfBk   
3703   4605  7M2tXmeS15NAzEn7ABFeBg   
2249   2542  1EaKU4dMbesXXd3BrLCtYG   
79       79  2LTlO3NuNVN70lp2ZbVswF   
777     817  1WFATT2cfUTj4Fbd0dJ9tq   
125     127  56ehF3tXMjwLJkAmuKzRE2   
2192   2461  5f9JYuHSIHjp6rggue13RS   
2013   2235  3RvLAm0ESbUOUZOD7sF4xn   
3487   4230  2aQwdBWk4vSPWk1Csgtzdo   
2674   3053  4Pwjz3DfvfQWV0rO2V8jyh   
2585   2939  3eMsgWDRq5dKl5SBTYGv2J   
2766   3191  439TlnnznSiBbQbgXiBqAd   
1018   1080  26v2Gf4Af7Fav0jTQluetL   
1575   1718  4svCVV4VHEL3Z4efES868h   
2767   3192  4SE4yewyGpOYfxfx59Yjc5   
3406   4110  5gFoAVTN9YlM9uJCrFZtgl   
3214   3866  2DLsQuwxEdXt2yp54vw355   
3315   4009  6PGoSes0D9eUDeeAafB2As   
3501   4249  40IQooNkPRVtn4zlE3ZFpA   
2759   3181  7KXjTSCq5nL1LoYtL7XAwS   
3348   4046  5cXg9IQS34FzLVdHhp7hu7   
3126   3731  75IN3CtuZwTHTnZvYM4qnJ   
3689   4583  0N3W5peJUQtI4eyR6GJT5O   
3496   4241  2Fw5S2gaOSZzdN5dFoC2dj   
3403   4107  3I0FBDc1c1BLNtXWKVjmFg   
3310   4002  7yotKA30dwTKNEGomV9ZsI   
2197   2466  0dbTQYW3Ad1FTzIA9t90E8   
2386   2710  4hsncV3bzuB83dBJcnSCBl   
2765   3189  6BCs8vxnGgKL9eMCOrG2wm   
2763   3186  3ZpwwnyXZLppRzSY1DG5lx   
121     121  5gmkw2031VVIboYqPIjSWB   
3313   4007  2JvzF1RMd7lE3KmFlsyZD8   
2789   3218  5eEj1vkMIqtmvaTk7cS4UC   
3691   4586  4fzsfWzRhPawzqhX8Qt9F3   
3529   4296  7JuB63pWkmxIxSazNB6F5m   
124     124  42zd6DYQ4o4SECmTITrM1U   
2760   3182  5rxPi0MiR4miNK0rCmk980   
2084   2322  5YGhm6pI5s0uxJ2DwrDYqO   
3638   4491  3iVcZ5G6tvkXZkZKlMpIUs   
3591   4404  0utlOiJy2weVl9WTkcEWHy   
621     641  5BSz8sJO2YmKfNVQG7fq2J   
2668   3046  6Ius4TC0L3cN74HT7ENE6e   
2790   3219  6VDz4m1vEGHDDqQojf1mpc   

                                                   name  popularity  \
3121                                 Backseat Freestyle          70   
2966                                               DNA.          80   
4010                                                ATM          67   
2976                                        Mockingbird          78   
3309                                         DUCKWORTH.          69   
2191                     Not Alike (feat. Royce Da 5'9)          69   
2200                                   Never Love Again          67   
3218                                             Uproar          71   
3703                            Homicide (feat. Eminem)          76   
2249                                           ELEMENT.          73   
79                                                FEEL.          67   
777                                           Amorphous          28   
125                                           Surf Swag          56   
2192                                           Mahogany          56   
2013                                        Case Closed          72   
3487                                          This Life          61   
2674                          Bitch, Don’t Kill My Vibe          69   
2585                                             Higher          74   
2766                                       m.A.A.d city          69   
1018                                           Penguins          66   
1575                                              Go DJ          61   
2767                                     Cinderella Man          71   
3406                          Crooked Smile (feat. TLC)          72   
3214                                         No Worries          60   
3315                                LOVE. FEAT. ZACARI.          81   
3501                          untitled 07 | 2014 - 2016          60   
2759                                            HUMBLE.          84   
3348                       B

In [37]:
#Checking to see how many songs have a probability of me enjoying it over .75
new_tracks = recommendation_tracks_df[recommendation_tracks_df['prob_ratings'] >= 0.75]
print(len(new_tracks))

116


Success! The song recommendations looks like straight fire, and I am excited to add some of these tracks to my playlists. There are 116 songs that have a probability rating of over 0.75, which I think is pretty good.

### Some takeaways from the data:

-The first thing I notice is that the song recommendations are very similar to what I listen to on a regular basis. As I predicted, because I used my top artist data to rate the songs, the model recommends a lot of songs by these same artists. However, there are several different artists who are not in my top artists who I still recognize and enjoy, as well as some artists I have never heard of. 

-My top artists dominate my song recommendations. Perhaps if I used specific song data as well as top artist data I would get a bit more diversity in the recommendations, but in doing that it may also reduce the probability ratings of the songs suggested.

In [38]:
table = classification_report(rfc.predict(X_test), y_test)

In [50]:
print(tabulate(top_tracks_df.name, headers='firstrow', tablefmt='fancy_grid'))

╒════╤════╤════╤════╤════╤═════╤═════╤═════╤═════╤═════╤═════╤═════╤═════╤═════╤═════╤═════╤═════╤═════╤═════╕
│    │    │    │    │    │ 0   │ L   │ e   │ s   │     │ V   │ a   │ l   │ s   │ e   │ u   │ s   │ e   │ s   │
╞════╪════╪════╪════╪════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╡
│  1 │ Y  │ e  │ b  │ b  │ a   │ ’   │ s   │     │ H   │ e   │ a   │ r   │ t   │ b   │ r   │ e   │ a   │ k   │
├────┼────┼────┼────┼────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┤
│  2 │ C  │ a  │ m  │ e  │     │ F   │ o   │ r   │     │ T   │ h   │ e   │     │ L   │ o   │ w   │     │     │
├────┼────┼────┼────┼────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┤
│  3 │ R  │ E  │ A  │ D  │ Y   │     │ T   │ O   │     │ G   │ O   │     │     │     │     │     │     │     │
├────┼────┼────┼────┼────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┤
│

In [40]:
def classification_report_csv(report):
    report_data = []
    lines = report.split('\n')
    for line in lines[2:-3]:
        row = {}
        row_data = line.split('      ')
        row['class'] = row_data[0]
        row['precision'] = float(row_data[1])
        row['recall'] = float(row_data[2])
        row['f1_score'] = float(row_data[3])
        row['support'] = float(row_data[4])
        report_data.append(row)
    dataframe = pd.DataFrame.from_dict(report_data)
    dataframe.to_csv('classification_report.csv', index = False)

In [42]:
report = classification_report(rfc.predict(X_test), y_test)
classification_report_csv(report)

IndexError: list index out of range

In [54]:
top_tracks_df.name.head(10)

0                                        Les Valseuses
1                                   Yebba’s Heartbreak
2                                     Came For The Low
3                                          READY TO GO
4                                     Don't Wanna Know
5                                SUVs (Black on Black)
6    Who Want Smoke?? (feat. G Herbo, Lil Durk & 21...
7    JU$T (feat. Pharrell Williams & Zack de la Roc...
8                                         Garden Party
9                                       Still Trappin'
Name: name, dtype: object